### ISSUE TITLE
Add ability to limit max repository size.

### ISSUE THREAD

**Issue by slynn1324** [restic/restic#4585]
Output of `restic version`
--------------------------

restic 0.16.2 compiled with go1.21.3 on linux/amd64

What should restic do differently? Which functionality do you think we should add?
----------------------------------------------------------------------------------

Add the ability to limit the maximum repository size.  

What are you trying to do? What problem would this solve?
---------------------------------------------------------

Some cloud storage providers (Wasabi) don't seem to provide a way to limit the size of a bucket or a spending limit on the account.  Being able to set a maximum-size threshold on the backup repository will help prevent a possible expensive accidental overrun if it begins backing up more than expected due to some unexpected condition such as passing the wrong directory arguments or moving more files into a directory than possibly intended for the backup set.  

I currently can approximate this in a wrapper script by first running `restic stats --json mode raw-data`, followed by `restic backup --quiet --dry-run --json` and adding up the 'total_size' from the stats call and 'data_added' from the backup dry run to get a new projected total size -- and checking that against a configured threshold.  

It would be nice (and I imagine possibly more efficient?) to be able to pass a "--max-repo-size 50G" argument to restic directly. 

If I'm missing some other way of accomplishing this that exists -- my apologies and please advise.  Thanks!! 

Did restic help you today? Did it make you happy in any way?
------------------------------------------------------------

Restic is awesome!  I love the ways that it lets me manage backups in an efficient way across multiple different back ends, and is supported across all operating systems!  Great work!

**Comment by dia3olik** [comment]
Yeah! Any news on this?

**Comment by MichaelEischer** [comment]
There is one key limitation for this feature. It won't be possible to accurately limit the repository if multiple clients backup to the same repository at the same time. There will also be a significant slowdown on startup as restic would have to list all files in the backend to know how large the repository currently is.

**Comment by wplapper** [comment]
A quick estimate of the repository size is the sum of sizes of all packfiles, excluding space needed for indices and snapshots etc.:
 repo_packs, _ := pack.Size(ctx, repo, false)
  pack_size := int64(0)
  for _, size := range repo_packs {
    pack_size += size
  }
This is an index function and will not go back to the backend.

IMHO one doesn't need an exact limit, just a rough limit. If multiple clients are backing up at the same time, the last one hitting the limit check will have to bite the bullet (and the backup will fail).

**Comment by MichaelEischer** [comment]
> This is an index function and will not go back to the backend.

True, that is also already used to clean up stale pack files from the cache. Judging from the described use case, a rough limit would be sufficient. Together with the size of the index files that should cover 99% of the repository size.

What would be the expected behavior if the repository size is reached? Then restic should probably just cancel the backup creation?

**Comment by wplapper** [comment]
I think cancelling the backup would be the right choice, together with an appropriate error message. Then the administrator / user can take manual steps to either increase the repository size or remove old snapshot (and the prune it).

**Comment by wplapper** [comment]
@MichaelEischer: I have a first draft of how to implement this feature. I know how to detect the condition, when the repo size is exceeded, but currently I leave a lot of dead bodies = unreferenced data packfiles lying around. 
I can deal with that by pruning them out of the repository, but they are not seen by pack.Size(ctx, repo, false) since they are not referenced. Only prune looks for them and then removes them.
So multiple (failed) backup commands still increase the real size of the repository. 

I would like to cleanly shutdown the backup process by saving the failed backup as a new snapshot with a tag "size-exceeded" or similar, so the data is visible and accounted for and can be managed properly by forget / prune..

I will upload the draft soon.

**Comment by smoothscholar** [comment]
@wplapper is there something i am missing here? as i understand it, right now restic will fill the drive its given completely, with no option to limit size, and no option to automatically delete snapshots

once it fills drive, the backup will fail, no matter what.

is my understanding wrong? if its right, why is this not a pressing issue?

**Comment by wplapper** [comment]
> @wplapper is there something i am missing here? as i understand it, right now restic will fill the drive its given completely, with no option to limit size, and no option to automatically delete snapshots.

Yes, ``restic backup`` will fill the drive or the quota for the drive with no limit, assuming you produce new data for every new backup run. 

But restic is meant to be a command line tool and not a fully automated application with all the bells and whistles and a GUI. Therefore you have other commands like ``restic forget`` and ``restic prune`` which will delete snapshots and keep the repository in shape. ``restic forget`` allows you to set up policies which will automatically delete snapshots when ``forget`` is run.

With a bit of scripting or other tools which use restic as their hard worker one can generate a backup solution which will maintain the repositories and will look after them with minimal additional intervention. 

> is my understanding wrong? if its right, why is this not a pressing issue? 

Yes it is a pressing issue for some users like @slynn1324, but the maintainers of restic might have a different view. See https://github.com/restic/restic/issues/21793 for another discussion around this subject. I am just a mere contributor to restic.

**Comment by MichaelEischer** [comment]
> as i understand it, right now restic will fill the drive its given completely, with no option to limit size, and no option to automatically delete snapshots
> 
> once it fills drive, the backup will fail, no matter what.

Deleting snapshots when running out of space implicitly also require deciding _which snapshot_ to delete. This is something only the user can decide or some automatic scripting around restic.

In terms of core backup functionality, the part relevant for inclusion in restic would be a way to prevent backups from exceeding a certain size. Although that cannot be 100% reliable as multiple restic processes that don't know much of each others activities can back up to the same repository. But it might still be useful for certain use cases.
